File: 7_Sets/javascript/HashedSetDataStructure.js
export class MyHashedSet {
  constructor() {
    this.hashSet = new Map();
  }

  // Note: this is intentionally simplistic for teaching.
  // For objects, JSON.stringify can be unstable if key ordering differs.
  hash(value) {
    const t = typeof value;
    if (value === null) return "null";
    if (t === "string") return `s:${value}`;
    if (t === "number") return `n:${Object.is(value, -0) ? "-0" : String(value)}`;
    if (t === "boolean") return `b:${value}`;
    return `j:${JSON.stringify(value)}`;
  }

  add(value) {
    this.hashSet.set(this.hash(value), value);
  }

  remove(value) {
    this.hashSet.delete(this.hash(value));
  }

  contains(value) {
    return this.hashSet.has(this.hash(value));
  }

  [Symbol.iterator]() {
    return this.hashSet.values();
  }

  toString() {
    const items = [...this.hashSet.values()];
    if (items.length === 0) return "set()";
    return `{${items.map(String).join(", ")}}`;
  }
}

File: 7_Sets/python/HashedSetDataStructure.py
class MyHashedSet:
    def __init__(self):
        self.hashSet = {}
    
    def add(self, value):
        self.hashSet[hash(value)] = value
    
    def remove(self, value):
        if hash(value) in self.hashSet:
            del self.hashSet[hash(value)]
    
    def contains(self, value):
        return hash(value) in self.hashSet
    
    def __iter__(self):
        for i in self.hashSet.values():
            yield i
        return None
    
    def __str__(self):
        if len(self.hashSet) == 0:
            return "set()"
        else:
            return "{" + ", ".join([str(i) for i in self.hashSet.values()]) + "}"